[TF:TRT] Add Grappler optimizer that executes rewrites for TF-TRT conversion preperation

christopherbate opened the pull request:
This change adds a grappler pass prior to the TRTOptimizer. The new pass runs rewrites on the GraphDef via fixed point iteration. Some of the code is borrowed from the arithmetic optimizer. The TF-TRT segmentation and conversion process may desire to rewrite certain operation patterns in order to facilitate conversion or the performance of the converted model. Currently, we only perform one rewrite on the GraphDef before converting (FP32 cast rewrite). This ad-hoc rewrite is reformulated into a formal rewrite stage of the grappler pass. Additionally, this changes adds a rewriter to recognize when all required shape information is present for Conv2DBackpropInput and adds the information required for TRT conversion into a special attribute.

christopherbate commented:
I think we'll see the use cases for this will keep growing. I added a commit `Add conditioning pass to enable Conv2Dbackprop const fold WAR`. This resolves an existing issue for `Conv2DBackpropInput` converter when the input has dynamic batch size. I can split this last commit off into a separate PR , but currently I am including it here currently to highlight the use case.

The issue is that TF's Conv2dBackpropInput op requires a parameter `input_sizes` which is actually the full shape of the output of the Conv2DBackpropInput op instance, including batch dimension and channel dimensions. Operations dealing with a dynamic batch dimension cannot be const-folded by Grappler.

User's code typically looks like this:

```
batch_size = tf.shape(input_tensor)[0] # this StridedSlice cannot be const-folded
non_batch_dims = tf.shape(input_tensor)[1:]   # This can and will be const-folded by Grappler's default const-folding optimizer
output_shape = tf.stack([batch_size, non_batch_dims]) # Thus, this Stack cannot be const-folded
upsampled_tensor = tf.conv2d_backprop_input(input_tensor, input_sizes=output_shape, ...)
```

The net result is that "output_shape" is a non-constant tensor.

However, for TF-TRT conversion, we don't care/use about the non-spatial dimensions of the `input_sizes` input tensor, so to us this looks like we should be able to const-fold the graph if we ignore the batch dimension.

The commit adds a custom rewriter to recognize when Conv2DBackpropInput has the situation shown above, including that we only have dynamic batch dimension above, and all other dimensions are static. 
Since we need to preserve originally graph def for native execution, we find the relevant shape constants and assign as a custom attribute to the`Conv2DBackpropInput` `NodeDef` instance under consideration. During conversion time, we first check for this attribute. If it exists, then we ignore the "input_sizes/output_shape" op input entirely, and use the attribute parameters to achieve conversion.

bixia1 commented:
I see this PR in draft state.  We have discussed this at the meeting and agreed on this approach.
Are you able to split this into a sequence of a few PRs? It looks to me most individual commits here can be a separated PRs.

christopherbate commented:
First commit was split out into [this PR](https://github.com/tensorflow/tensorflow/pull/53652), which has all comments addressed.

christopherbate commented:
@bixia1 The second commit is now pulled out as #54301

gbaned commented:
@christopherbate This PR is in draft, any update on this? Please. Thanks!

christopherbate commented:
I've been pulling individual commits out as seperate PRs.  The last one, which is still awaiting merge, is #54301

christopherbate commented:
@bixia1 I've combined the remaining changes into a single commit, and this is ready for review.

There are additional rewrites for cast that I've removed and will introduce in a follow-on PR.

bixia1 commented on tensorflow/compiler/tf2tensorrt/convert/conditioning_pass.cc:
```diff
@@ -0,0 +1,401 @@
+
+namespace {
+
+// Extract values from a Const op to `values`. Returns true if succeeds.
```
s/Extract/Extracts/

bixia1 reviewed (commented):
Would you please split this PR into two:
(1) add framework and move the existing FP32 cast  conversion to RewriteFP32CastsStage
(2) add FoldConv2DBackpropOutputShape to the pipeline

christopherbate committed:
[TF:TRT] Add preperation pass

christopherbate commented:
Rebased and addressed comments. As discussed this week, I removed parts specific to the Conv2DBackprop rewrite, which will follow after this is merged.

christopherbate commented:
@bixia1  any progress on review?

bixia1 commented:
> Rebased and addressed comments. As discussed this week, I removed parts specific to the Conv2DBackprop rewrite, which will follow after this is merged.

Would you please fix the PR description to reflect this change?

bixia1 commented on tensorflow/compiler/tf2tensorrt/convert/preperation_pass.cc:
```diff
@@ -0,0 +1,310 @@
+}
+
+static grappler::CustomGraphOptimizerRegistrar TRTPreperationPass_Registrar(
+    []() { return new TRTPreperationPass("TensorRTConditioner"); },
```
The name here is not consistent with what is used in trt_convert.py.
How do you test this PR?

gbaned commented:
Hi @christopherbate  Can you please check @bixia1's comments and keep us posted ? Thank you!
